vortex-io/src/object_store.rs

```diff
@@ -76,10 +77,59 @@ impl VortexReadAt for ObjectStoreReadAt {
 
         Box::pin(async move {
             let start_range = pos as usize;
-            let bytes = object_store
-                .get_range(&location, start_range..(start_range + len as usize))
+
+            // Allocate an aligned vector to read into.
+            //
+            // This is
+            let alloc_size =
+                ((len as usize) + (BUFFER_ALIGNMENT - 1)).next_multiple_of(BUFFER_ALIGNMENT);
+            let mut buf = Vec::<u8>::with_capacity(alloc_size as _);
+            let padding = buf.as_ptr().align_offset(BUFFER_ALIGNMENT);
+            unsafe { buf.set_len(padding) };
+
+            let get_range = start_range..(start_range + len as usize);
+
+            let response = object_store
+                .get_opts(
+                    &location,
+                    GetOptions {
+                        range: Some(GetRange::Bounded(get_range)),
+                        ..Default::default()
+                    },
+                )
                 .await?;
-            Ok(bytes)
+
+            let mut byte_stream = response.into_stream();
+            while let Some(bytes) = byte_stream.next().await {
+                let bytes = bytes?;
+                buf.extend_from_slice(&bytes);
+            }
```

Comment:
this is pretty much exactly how ObjectStore::get_range is implemented so there's no more copying than before

Reply:
they have a small optimisation where if there's only single element response they don't copy it

Reply:
That's fair. Though I can't think of a way to make the kernel write HTTP payloads directly into aligned memory, so I don't think you can do this truly zero copy if you're using HTTP(S)

Reply:
Perhaps it's possible with HTTP2/3 since those don't have the variable length text headers, haven't looked into it